support feature flag overrides via query string

christianvogt opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
https://issues.redhat.com/browse/RHOAIENG-7894

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

Once feature flags are overridden in the current session a banner is displayed at the top of the page. From the banner the user can restore their session without the need for closing their browser window. They can also open a modal to more easily adjust the feature flags for the current session.
![image](https://github.com/opendatahub-io/odh-dashboard/assets/14068621/3b909643-70c8-4f6f-986f-0c94a7d2006a)

The modal lists all available feature flags and their current state. Flags that have been overridden in the session via query string or through this model are annotated as `(overridden)`. If the current dashboard config doesn't contain a valid entry (likely only a dev issue with a mismatched frontend and backend), the checkbox will be in the partially checked state.
![image](https://github.com/opendatahub-io/odh-dashboard/assets/14068621/e15d4f2e-1ef3-49ef-899b-f9363d7e5b79)

Changes to feature flags are immediately reflected in the UI.

While the modal is open, the query string is updated for the purpose of constructing a shareable URL that can also be bookmarked for future use.

Restoring defaults will immediately remove the banner.

To enable this feature, set the query string param to `?devFeatureFlags`.
To adjust the feature flags via the query string, assign a value: `?devFeatureFlags=home,disableInfo,disableAppLauncher=false,enablement=true`
Note that the query string accepts both variants of a flag which has the `disable` prefix or not. eg. `disableHome=true` is the same as `home=false`, or `disableHome=false` is the same as `home` or `home=true`

Some backend end points such as `/api/cluster-settings` reads from the dashboard config cached on the backend. In order to get around this caching, if feature flags are overridden in the front end, a new header is sent along with all requests. See `x-odh-feature-flags`.

The new header is supplied by setting the default headers for axios. As such all usage of axios must use our new singleton instance from `~/utilities/axios`. To ensure compliance a new eslint rule is provided that errors if the node_modules axios instance is imported.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Open any page and then change the query string to be `?devFeatureFlags=...` where `...` is either empty or a feature flag of your choice. eg `home`

Play around with the banner and feature flag modal.

As an admin navigate to `Settings => Cluster settings` and inspect the network tab. Look for the `/api/cluster-settings` call and inspect the response such that the enablement matches the front end feature flag settings:
```
    "modelServingPlatformEnabled": {
        "kServe": true,
        "modelMesh": true
    }
```

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
Added unit tests.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

andrewballantyne commented on frontend/src/app/__tests__/DevFeatureFlagsBanner.spec.tsx:
```diff
@@ -0,0 +1,80 @@
+import DevFeatureFlagsBanner from '~/app/DevFeatureFlagsBanner';
+
+describe('DevFeatureFlagsBanner', () => {
+  it('should render not render if no feature flags are overridden', () => {
```
```suggestion
  it('should not render if no feature flags are overridden', () => {
```

andrewballantyne commented on frontend/src/app/useDevFeatureFlags.ts:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // update session storage
+    // compares against oroginal sessionFlags object on purpose and not the sanitizedSessionFlags
```
```suggestion
    // compares against original sessionFlags object on purpose and not the sanitizedSessionFlags
```

andrewballantyne reviewed (changes requested):
Some typos.

I'm a little weary of the reliance on Axios; I had hoped to get rid of it in time as `fetch` can do all the needed items and it is one less dependency. Also, it's going to be interesting how the data only gets to the DashboardConfig on the backend when it is through a request of the frontend -- this will have weird edge cases in upgrades and refreshing the cache of the DashboardConfig. But maybe it won't be too problematic.

This looks great, even if it has potentially some limitations. This is a huge step forward for testing. 

I guess the question is, is this problematic on fully prod clusters 🤔 And I'm thinking no... because the backend DSC flags can still be disabled to stop flows.

christianvogt commented on frontend/src/app/App.tsx:
```diff
@@ -115,6 +119,10 @@ const App: React.FC = () => {
             <ErrorBoundary>
+              <DevFeatureFlagsBanner
+                dashboardConfig={dashboardConfig.spec.dashboardConfig}
+                {...devFeatureFlags}
```
Renamed to `devFeatureFlagsProps`

christianvogt commented:
> this will have weird edge cases in upgrades and refreshing the cache of the DashboardConfig. But maybe it won't be too problematic.

This is strictly a means for testing so any limitations are ok. That being said, upgrades will always use the config that's found on cluster. A frontend with session dev flags will still pick up any config changes from the cluster so long as they haven't overridden those flags that are changing; the UI will update accordingly.

> I guess the question is, is this problematic on fully prod clusters 🤔 And I'm thinking no... because the backend DSC flags can still be disabled to stop flows.

@andrewballantyne can you elaborate on this question.

andrewballantyne commented:
> > this will have weird edge cases in upgrades and refreshing the cache of the DashboardConfig. But maybe it won't be too problematic.
> 
> This is strictly a means for testing so any limitations are ok. That being said, upgrades will always use the config that's found on cluster. A frontend with session dev flags will still pick up any config changes from the cluster so long as they haven't overridden those flags that are changing; the UI will update accordingly.

This is a good point. It just has gaps that won't be able to be used for verification but there should be alternatives to getting around that without too much concern. Thanks.

> > I guess the question is, is this problematic on fully prod clusters 🤔 And I'm thinking no... because the backend DSC flags can still be disabled to stop flows.
> 
> @andrewballantyne can you elaborate on this question.

Well, consider a public hosted OpenShift AI -- allowing users to flip things on may reveal functionality not intended for use and a feature flag is not able to stop it now. Most features are backed by an OpenShift AI backend component, so it being disabled should be enough to stop functionality. 

But if a backend component is available, and the UI is disabled -- this could allow for enabling features not intended for public consumption. Although, I'm not sure that has a huge problem 🤔 Or maybe a flow that is K8s as the backend... that could be problematic too as there is "no way" to turn it off.

andrewballantyne commented on frontend/src/app/useDevFeatureFlags.ts:
```diff
@@ -0,0 +1,134 @@
+
+/**
+ * Override dashboard config feature flags in the query string: eg.
+ * `devFeatureFlags=disableHome=false,appLauncher,support=true`
+ * Results in:
+ *  - disableHome = true
```
Oh... I wanna say if I read the code this is a documentation bug not a functionality bug.

```suggestion
 * `devFeatureFlags=disableHome=false,appLauncher,support=true`
 * Results in:
 *  - disableHome = false
```

christianvogt commented:
> But if a backend component is available, and the UI is disabled -- this could allow for enabling features not intended for public consumption. Although, I'm not sure that has a huge problem 🤔 Or maybe a flow that is K8s as the backend... that could be problematic too as there is "no way" to turn it off.

The only extra bit of functionality this enables is the ability to supply feature flags to backend points that happen to use them. Today it is already possible for someone to enable all features themselves with dev tools in the browser without the need for changing the dashboard config.

christianvogt commented:
@andrewballantyne Thanks for the review. Sorry but just I added an extra feature  :)

While the modal is open, the query param is now visible and updated in real time. This gives the ability to more easily construct the URL, bookmark it and share it. While the modal is closed the query string param is once again removed to best represent a normal user experience.

Do you think we should also include a `hide` banner option so that we can utilize this feature for screenshots without the extra clutter? I can see this as useful and hides the banner until the page refreshes.

mturley reviewed (approved):
LGTM, pulled locally and tested, this is really awesome stuff
